### ISSUE TITLE
`jax.numpy.asarray` does not function under `jax.make_jaxpr`

### ISSUE THREAD

**Issue by patrick-kidger** [google/jax#18020]
### Description

```python
>>> import jax
>>> jax.make_jaxpr(jax.numpy.asarray)(1.)
{ lambda ; a:f32[]. let  in (a,) }
```
This appears to be a variant of #15676.

### What jax/jaxlib version are you using?

JAX 0.4.18

**Comment by jakevdp** [comment]
I think this is working as expected: `jax.make_jaxpr` converts inputs to JAX tracers at its boundary, and so by the time the tracer hits `jnp.asarray` it is already a `jax.Array`, and so `jnp.asarray` is a no-op.

**Comment by patrick-kidger** [comment]
That's certainly the reason for it. I don't think it's desired behaviour though -- it means that the round-trip function->jaxpr->function is not consistent:
```python
import jax
jaxpr = jax.make_jaxpr(jax.numpy.asarray)(1.)
[out] = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, 1.)
print(type(out))  # float
```
FWIW I hit this whilst using it in this way; I was surprised to find that the `asarray` statements of my closure-converted function were being ignored.

**Comment by jakevdp** [comment]
Yeah, I think if you're using internal tools like `eval_jaxpr`, it's up to you to make sure the inputs match what the jaxpr expects. We could probably change `eval_jaxpr` so that it does this conversion automatically, just like `make_jaxpr` does. What do you think?

**Comment by patrick-kidger** [comment]
This arises using the public JAX API:
```python
import jax

def f(x):
    y = x**2
    return jax.numpy.asarray(x)

f, consts = jax.closure_convert(f, 1.)
out = f(1., *consts)
print(type(out))
```
I'm only discussing internals like `eval_jaxpr` because we're both familiar with them and it's more of a MWE than the public API provides.

Incidentally note that the above example is also typical of my use case, where `asarray` is used on the output of a function, not the input.

**Comment by jakevdp** [comment]
`closure_convert` abstractifies all arguments, just like `make_jaxpr`, `jit`, etc. So I think you need to ensure the inputs to the closure converted functions are arrays.

I'm not familiar enough with `closure_convert` to judge whether this is a bug or if it's user error.

For what it's worth, an easy fix would be to wrap `f` in `jax.jit` before calling it.

**Comment by patrick-kidger** [comment]
It's nothing to do with the inputs. Here's another example:
```python
import jax

def f():
    return jax.numpy.asarray(1)

f, consts = jax.closure_convert(f)
out = f(*consts)
print(type(out))  # int

```
This is a function which has observably different behaviour after closure conversion. Code like `out.shape` will fail after the closure-conversion line is inserted.

To add a bit more colour: in my typical use case, I have something like:
```python
# Provided by a user
def f(...):
    ...

# Library code
def g(...):
    out = f(...)
    return jax.tree_util.tree_map(jnp.asarray, out)

g = jax.closure_convert(g, ...)
```
where the purpose of `g` is explicitly to ensure that all outputs are cast to arrays, and the closure-conversion is causing this to be ignored.

**Comment by jakevdp** [comment]
Makes sense, thanks. It looks like the implementation of `closure_convert` is wrong in cases of bound constants.

**Comment by jakevdp** [comment]
I'll note though that the `jit` fix works in this case too:
```
out = jax.jit(f)(*consts)
print(type(out))  # <class 'jaxlib.xla_extension.ArrayImpl'>
```
